File: python/m1/gh_auth/github_agent.py
# python/m1/gh_auth/github_agent.py
"""GitHub MCP agent — OAuth via the MCP SDK's built-in flow.

The MCP SDK's OAuthClientProvider handles PKCE, discovery, and token exchange.
GitHubTokenStorage pre-seeds the registered client credentials so the SDK skips
dynamic client registration (which GitHub does not support).

Run:  uv run python m1/gh_auth/github_agent.py
"""

from __future__ import annotations

import asyncio
import os
import threading
import webbrowser
from http.server import BaseHTTPRequestHandler, HTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

import httpx
from deepagents import create_deep_agent
from dotenv import load_dotenv
from langchain_mcp_adapters.client import MultiServerMCPClient
from langchain_mcp_adapters.tools import load_mcp_tools
from mcp.client.auth import OAuthClientProvider
from mcp.shared.auth import OAuthClientInformationFull, OAuthClientMetadata, OAuthToken

from models import model

ENV_PATH = Path(__file__).resolve().parent.parent.parent / ".env"
load_dotenv(dotenv_path=ENV_PATH, override=True)

TOKEN_FILE = Path(__file__).resolve().parent.parent.parent / ".m1_github_token"

MCP_URL = "https://api.githubcopilot.com/mcp/"
REDIRECT_URI = "http://127.0.0.1:8765/"
SCOPE = "repo read:user"


class GitHubOAuthClientProvider(OAuthClientProvider):
    """OAuthClientProvider with two GitHub-specific fixes:

    1. Scope: the SDK replaces client_metadata.scope with the full list from
       GitHub's PRM scopes_supported. We restore the intended scope before the
       authorization URL is built.
    2. Token response: GitHub returns URL-encoded form data by default; adding
       Accept: application/json forces a JSON response the SDK can parse.
    """

    def __init__(self, *args, requested_scope: str, **kwargs) -> None:
        super().__init__(*args, **kwargs)
        self._requested_scope = requested_scope

    async def _perform_authorization_code_grant(self) -> tuple[str, str]:
        self.context.client_metadata.scope = self._requested_scope
        return await super()._perform_authorization_code_grant()

    async def _exchange_token_authorization_code(
        self, auth_code: str, code_verifier: str, *, token_data: dict | None = {}
    ) -> httpx.Request:
        req = await super()._exchange_token_authorization_code(
            auth_code, code_verifier, token_data=token_data
        )
        req.headers["Accept"] = "application/json"
        return req


class GitHubTokenStorage:
    """Pre-seeds client credentials so the SDK skips dynamic client registration."""

    def __init__(self, client_id: str, client_secret: str) -> None:
        self._client_id = client_id
        self._client_secret = client_secret

    async def get_tokens(self) -> OAuthToken | None:
        if TOKEN_FILE.exists():
            raw = TOKEN_FILE.read_text().strip()
            if raw:
                try:
                    return OAuthToken.model_validate_json(raw)
                except Exception:
                    return OAuthToken(access_token=raw, token_type="Bearer")
        return None

    async def set_tokens(self, tokens: OAuthToken) -> None:
        TOKEN_FILE.write_text(tokens.model_dump_json())

    async def get_client_info(self) -> OAuthClientInformationFull:
        return OAuthClientInformationFull(
            client_id=self._client_id,
            client_secret=self._client_secret,
            redirect_uris=[REDIRECT_URI],
            token_endpoint_auth_method="client_secret_post",
        )

    async def set_client_info(self, info: OAuthClientInformationFull) -> None:
        pass  # credentials come from .env, not from registration


async def open_browser(url: str) -> None:
    print(f"Opening browser for GitHub login ({SCOPE})...")
    print(f"If it doesn't open, paste this URL:\n{url}\n")
    webbrowser.open(url)


async def wait_for_callback() -> tuple[str, str | None]:
    """Loopback server that captures the authorization code from GitHub's redirect."""
    result: dict[str, str] = {}
    done = threading.Event()

    class Handler(BaseHTTPRequestHandler):
        def do_GET(self):  # noqa: N802
            params = parse_qs(urlparse(self.path).query)
            result["code"] = params.get("code", [""])[0]
            result["state"] = params.get("state", [""])[0]
            self.send_response(200)
            self.send_header("Content-Type", "text/html")
            self.end_headers()
            self.wfile.write(b"<h3>Authorized. You can close this tab.</h3>")
            done.set()

        def log_message(self, *a): pass

    port = int(urlparse(REDIRECT_URI).port or 80)
    server = HTTPServer(("127.0.0.1", port), Handler)
    threading.Thread(target=server.serve_forever, daemon=True).start()

    if not done.wait(timeout=300):
        server.shutdown()
        raise SystemExit("Timed out waiting for OAuth callback.")
    server.shutdown()
    return result["code"], result.get("state")


async def main() -> None:
    client_id = os.environ.get("GITHUB_CLIENT_ID")
    client_secret = os.environ.get("GITHUB_CLIENT_SECRET")
    if not client_id or not client_secret:
        raise SystemExit("Set GITHUB_CLIENT_ID and GITHUB_CLIENT_SECRET in .env.")

    auth = GitHubOAuthClientProvider(
        server_url=MCP_URL,
        client_metadata=OAuthClientMetadata(
            redirect_uris=[REDIRECT_URI],
            scope=SCOPE,
            token_endpoint_auth_method="client_secret_post",
        ),
        storage=GitHubTokenStorage(client_id, client_secret),
        redirect_handler=open_browser,
        callback_handler=wait_for_callback,
        requested_scope=SCOPE,
    )

    client = MultiServerMCPClient({
        "github": {
            "transport": "streamable_http",
            "url": MCP_URL,
            "auth": auth,
        }
    })

    async with client.session("github") as session:
        tools = await load_mcp_tools(session)
        for tool in tools:
            # BaseTool.handle_tool_error only catches ToolException, not the MCP
            # server's own errors (e.g. McpError on a 404), which would otherwise
            # crash the whole run. Wrap the coroutine so failures become a normal
            # tool result the agent can see and react to.
            def _make_safe(coro):
                async def _safe(*args, **kwargs):
                    try:
                        return await coro(*args, **kwargs)
                    except Exception as e:  # noqa: BLE001
                        # response_format="content_and_artifact" expects a 2-tuple
                        return f"Tool call failed: {e}", None
                return _safe

            tool.coroutine = _make_safe(tool.coroutine)
        print(f"github: {len(tools)} tool(s) available")

        agent = create_deep_agent(model=model, tools=tools)
        print("Running agent...\n")
        result = await agent.ainvoke({
            "messages": [{"role": "user", "content": (
                "What is my GitHub username and what repositories do I have?"
            )}]
        })
        print("\n" + result["messages"][-1].content)


if __name__ == "__main__":
    asyncio.run(main())

File: typescript/m1/gh_auth/github_agent.ts
// typescript/m1/gh_auth/github_agent.ts
/**
 * GitHub MCP agent — OAuth via the MCP SDK's OAuthClientProvider interface.
 *
 * GitHubTokenStorage implements OAuthClientProvider directly. It pre-seeds
 * the registered client credentials so the SDK skips dynamic client
 * registration (which GitHub does not support), and drives the
 * browser-consent + loopback-callback dance itself inside
 * redirectToAuthorization.
 *
 * Run:  pnpm tsx ./m1/gh_auth/github_agent.ts
 */

import { exec } from "node:child_process";
import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { createServer } from "node:http";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { config as loadDotenv } from "dotenv";
import { createDeepAgent } from "deepagents";
import { MultiServerMCPClient } from "@langchain/mcp-adapters";
import { auth } from "@modelcontextprotocol/sdk/client/auth.js";
import type { OAuthClientProvider } from "@modelcontextprotocol/sdk/client/auth.js";
import type {
  OAuthClientInformationFull,
  OAuthClientMetadata,
  OAuthTokens,
} from "@modelcontextprotocol/sdk/shared/auth.js";

import { model } from "../../models.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const ENV_PATH = path.resolve(HERE, "../../.env");
loadDotenv({ path: ENV_PATH, override: true });

const TOKEN_FILE = path.resolve(HERE, "../../.m1_github_token");

const MCP_URL = "https://api.githubcopilot.com/mcp/";
const REDIRECT_URI = "http://127.0.0.1:8765/";
const SCOPE = "repo read:user";

function openBrowser(url: string): void {
  const cmd =
    process.platform === "darwin" ? "open" : process.platform === "win32" ? "start" : "xdg-open";
  exec(`${cmd} "${url}"`);
}

class GitHubTokenStorage implements OAuthClientProvider {
  private codeVerifierValue?: string;
  private pendingCode?: string;

  constructor(
    private clientId: string,
    private clientSecret: string
  ) {}

  get redirectUrl(): string {
    return REDIRECT_URI;
  }

  get clientMetadata(): OAuthClientMetadata {
    return {
      redirect_uris: [REDIRECT_URI],
      scope: SCOPE,
      token_endpoint_auth_method: "client_secret_post",
    };
  }

  clientInformation(): OAuthClientInformationFull {
    return {
      client_id: this.clientId,
      client_secret: this.clientSecret,
      redirect_uris: [REDIRECT_URI],
      token_endpoint_auth_method: "client_secret_post",
    };
  }

  saveClientInformation(): void {
    // GitHub is pre-registered (see clientInformation); nothing to persist.
  }

  tokens(): OAuthTokens | undefined {
    if (!existsSync(TOKEN_FILE)) return undefined;
    const raw = readFileSync(TOKEN_FILE, "utf8").trim();
    if (!raw) return undefined;
    try {
      return JSON.parse(raw);
    } catch {
      return { access_token: raw, token_type: "Bearer" };
    }
  }

  saveTokens(tokens: OAuthTokens): void {
    writeFileSync(TOKEN_FILE, JSON.stringify(tokens));
  }

  saveCodeVerifier(codeVerifier: string): void {
    this.codeVerifierValue = codeVerifier;
  }

  codeVerifier(): string {
    if (!this.codeVerifierValue) throw new Error("No code verifier saved.");
    return this.codeVerifierValue;
  }

  /** Take (and clear) the authorization code captured by the last redirectToAuthorization call. */
  takePendingCode(): string | undefined {
    const code = this.pendingCode;
    this.pendingCode = undefined;
    return code;
  }

  async redirectToAuthorization(authorizationUrl: URL): Promise<void> {
    const expectedState = authorizationUrl.searchParams.get("state");

    console.log(`Opening browser for GitHub login (${SCOPE})...`);
    console.log(`If it doesn't open, paste this URL:\n${authorizationUrl}\n`);

    const code = await new Promise<string>((resolve, reject) => {
      const server = createServer((req, res) => {
        const params = new URL(req.url ?? "/", REDIRECT_URI).searchParams;
        if (params.get("state") !== expectedState) {
          res.writeHead(400);
          res.end("State mismatch. Close this tab and re-run.");
          return;
        }
        res.writeHead(200, { "Content-Type": "text/html" });
        res.end("<h3>Authorized. You can close this tab.</h3>");

        const authCode = params.get("code") ?? "";
        const error = params.get("error") ?? "";
        server.close();
        if (error || !authCode) {
          reject(new Error(`Consent failed: ${error || "no code returned"}`));
          return;
        }
        resolve(authCode);
      });

      const timeout = setTimeout(() => {
        server.close();
        reject(new Error("Timed out waiting for consent."));
      }, 300_000);
      timeout.unref();

      server.listen(8765, "127.0.0.1", () => openBrowser(authorizationUrl.toString()));
    });

    this.pendingCode = code;
  }
}

async function main(): Promise<void> {
  const clientId = process.env.GITHUB_CLIENT_ID;
  const clientSecret = process.env.GITHUB_CLIENT_SECRET;
  if (!clientId || !clientSecret) {
    throw new Error("Set GITHUB_CLIENT_ID and GITHUB_CLIENT_SECRET in .env.");
  }

  const authProvider = new GitHubTokenStorage(clientId, clientSecret);

  function makeClient() {
    return new MultiServerMCPClient({
      github: { transport: "http", url: MCP_URL, authProvider },
    });
  }

  let client = makeClient();
  try {
    let tools;
    try {
      tools = await client.getTools();
    } catch (err) {
      // First run: connect() throws once redirectToAuthorization captures a code,
      // because the SDK's auth() orchestrator always reports "REDIRECT" on the
      // attempt that triggered the browser flow, even though our own
      // redirectToAuthorization already has the code in hand by the time it
      // resolves. Feed that code back through auth() to do the actual token
      // exchange, then reconnect on a fresh client — the failed one can't retry.
      const code = authProvider.takePendingCode();
      if (!code) throw err;

      const result = await auth(authProvider, { serverUrl: MCP_URL, authorizationCode: code });
      if (result !== "AUTHORIZED") throw err;

      await client.close();
      client = makeClient();
      tools = await client.getTools();
    }

    for (const mcpTool of tools) {
      // DynamicStructuredTool.func only surfaces exceptions the SDK itself throws,
      // not the MCP server's own errors (e.g. an McpError on a 404), which would
      // otherwise crash the whole run. Wrap it so failures become a normal tool
      // result the agent can see and react to.
      const originalFunc = mcpTool.func;
      mcpTool.func = async (...args: Parameters<typeof originalFunc>) => {
        try {
          return await originalFunc(...args);
        } catch (err) {
          // responseFormat: "content_and_artifact" expects a 2-tuple
          return [`Tool call failed: ${err instanceof Error ? err.message : err}`, undefined];
        }
      };
    }

    console.log(`github: ${tools.length} tool(s) available`);

    const agent = createDeepAgent({ model, tools });
    console.log("Running agent...\n");
    const result = await agent.invoke({
      messages: [
        {
          role: "user",
          content: "What is my GitHub username and what repositories do I have?",
        },
      ],
    });
    console.log(`\n${result.messages.at(-1)?.content}`);
  } finally {
    await client.close();
  }
}

main().catch((exc) => {
  console.error(exc instanceof Error ? exc.message : exc);
  process.exit(1);
});
